src/lib.rs

```diff
@@ -78,95 +80,38 @@ impl<R> Volatile<R> {
     /// volatile.write(1);
     /// assert_eq!(volatile.read(), 1);
     /// ```
-    pub const fn new(reference: R) -> Volatile<R> {
+    pub unsafe fn new<A>(reference: &UnsafeCell<T>, access: A) -> Volatile<'_, T, A> {
         Volatile {
             reference,
             access: PhantomData,
         }
     }
 
-    /// Constructs a new read-only volatile instance wrapping the given reference.
-    ///
-    /// This is equivalent to the `new` function with the difference that the returned
-    /// `Volatile` instance does not permit write operations. This is for example useful
-    /// with memory-mapped hardware registers that are defined as read-only by the hardware.
-    ///
-    /// ## Example
-    ///
-    /// Reading is allowed:
-    ///
-    /// ```rust
-    /// use volatile::Volatile;
-    ///
-    /// let value = 0u32;
-    ///
-    /// let volatile = Volatile::new_read_only(&value);
-    /// assert_eq!(volatile.read(), 0);
-    /// ```
-    ///
-    /// But writing is not:
-    ///
-    /// ```compile_fail
-    /// use volatile::Volatile;
-    ///
-    /// let mut value = 0u32;
-    ///
-    /// let mut volatile = Volatile::new_read_only(&mut value);
-    /// volatile.write(1);
-    /// //ERROR: ^^^^^ the trait `volatile::access::Writable` is not implemented
-    /// //             for `volatile::access::ReadOnly`
-    /// ```
-    pub const fn new_read_only(reference: R) -> Volatile<R, ReadOnly> {
-        Volatile {
-            reference,
-            access: PhantomData,
-        }
+    pub fn from_ref(reference: &T) -> Volatile<'_, T, ReadOnly> {
```

Comment:
I don't think this and `from_mut_ref` are sound, but I am not quite sure.

Reply:
Hmm, I'm not sure about this either. I thought the &T/&mut T references are no longer live while the `Volatile` is active so that we don't break any aliasing guarantees. But as I said, I'm not sure about this...

Reply:
(no longer applies to the latest version)